feat: bahis 3.0 :tada:

ChasNelson1990 opened the pull request:
**DO NOT MERGE - THIS PR IS NOT FULLY FUNCTIONAL**

## Description

Features:

- [x] Reset database from File Menu
- [x] Sign in with three routes
  - [x] local (i.e. offline first)
  - [x] first time user (i.e. online first)
  - [x] change user (also online first)
- [x] Automatic app data sync on sign in (if online; tries no matter what sign in route is taken)
  - [x] Manual app data sync via File menu
- [x] Electron autoupdater
  - [x] Can be manually triggered via File menu (see TODO)
- [x] Menus: Containers, lists, forms and iFrames
- [x] Forms
  - [x] Empty form (with cloudUser and cloudCSV data)
  - [x] Fill out and save the form as draft
  - [x] Submit the form to KoboToolbox
  - [x] cloudUser tags replace user info
  - [x] cloudCSV Administrative Regions tags replace geolocation based on the local db (i.e. catchment only)
  - [x] cloudCSV Taxonomy tags replace with local db taxonomies (no autocomplete for select_multiple see: enketo/enketo#1301)
- [x] Sync, i.e. submit, local draft forms
  - [x] See a local unsynced (draft) count (see TODO)
- [x] Draft, i.e. locally saved draft form list
- [x] Cloud, i.e. synced, form list
- [x] iFrame for Dashboard

New features compared to v2:

- Modules can now have icons synced from bahis-management
- Massive upgrades of all dependencies (and removing mPower and other limiting deps)
- New styling engine via muiv5
- Kobo form styling (note: this actually influences the whole app but works quite nicely)
- Simplified code
- Fewer components (which all did the same stuff)
- No need for Reduce (just using the React app's own state)
- Form lists no longer need list definitions as we use the mui Datagrid element to provide column controls and table filtering

Deprecations compared to v2:

- Translations - I have currently removed all app language controls; these should be very easy to reintroduce in a later version
- Flat tables - I don't think we need flat tables for bahis-desk TBH, using the long-form table that is native to KoboToolbox and similar solutions works fine and is pretty speedy (for n<1000 at least)

TODO:

- [x] MAJOR: We can't edit locally saved draft forms (see enketo/enketo#1302)
- [x] MAJOR: Workflows don't work for the same reason
- [x] MAJOR: And we can't open synced forms for the same reason
- [x] Delete locally saved draft
- [ ] Manually triggered electron updater doesn't feed a message back to the user
- [ ] The local draft counter is quite slow to update (some stored state I think)
- [ ] What's new should be a dismissable modal rather than the accordion - and probably DB based rather than manually typed.
- [ ] Should we migrate v2 form data into this app somehow?
- [ ] Add username, app version and last sync time to API calls
- [ ] Add pagination to submissions API calls?
- [ ] Move over to BAHIS 3 user auth and roles
- [x] Check our get / post forms code handles redeployed forms in KoboToolbox? Do we need to store the specific deployment UID for submission?
- [x] There is a bug where if you try to view an existing form and then go back to the form list page it breaks - this might magically solve itself if the above major issues are solved.

## Dev Note

For some reason, on my linux box, I had to manually edit package.json for leaflet.googlemutant (bahis-desk/node_modules/leaflet.gridlayer.googlemutant/package.json) and change 'browser' to 'main' and delete the 'module' entry.

## Dependency Changes

Tonnes!

## Database Changes

Tonnes!

## Checklist

- [x] I have read the road86 Contribution Guide.
- [x] I have checked all commit message styles match the requested structure.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have assigned at least one reviewer.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] My changes generate no new warnings.
- [ ] I have made corresponding changes to the documentation.
- [x] New database changes have been committed.

ChasNelson1990 committed:
feat: extract BAHIS2 server URL into env files

ChasNelson1990 committed:
fix: move to proper better-sqlite3 now the destyk PR has been released

ChasNelson1990 committed:
feat: add migration for old db location

ChasNelson1990 committed:
feat: add migration for old bahis data location to new location (can be deprecated with v3.0)

ChasNelson1990 committed:
feat: completely refactor remote GET / POST and local CRUD for BAHIS2 syncs

ChasNelson1990 committed:
feat: update README with the fact that everything for BAHIS2 is now explicitly "2"

ChasNelson1990 committed:
fix: lint

ChasNelson1990 committed:
feat: start initialising bahis 3 tables in local DB

ChasNelson1990 committed:
feat: get BAHIS 3 module definitions electron-side

ChasNelson1990 committed:
feat: sanitize error reporting in front end and make other minor changes

ChasNelson1990 committed:
feat: make react work with bahis 3 module definitions

ChasNelson1990 committed:
fix: santize styling for form details and list profiles

ChasNelson1990 committed:
fix: start stripping app language out

ChasNelson1990 committed:
feat: add material ui datagrid

ChasNelson1990 committed:
feat: rebuild lists so that they don't need list definitions anymore

ChasNelson1990 committed:
feat: remove unused sql set up script

ChasNelson1990 committed:
feat: add enketo-core

ChasNelson1990 committed:
fix: make git ignore local dev taxonomy files

ChasNelson1990 committed:
feat: split environment variables for bahis2 and 3 server (in default env so far)

ChasNelson1990 committed:
feat: bring redux MenuItem in line with bahis 3 DB model

ChasNelson1990 committed:
feat: add bahis 3 workflows, forms, taxonomies and admin regions to the local DB and sync

ChasNelson1990 committed:
feat: bring menus inline with BAHIS 3 model

ChasNelson1990 committed:
feat: correct form IDs to TEXT in local DB

ChasNelson1990 committed:
fix: make parent_administrative_region NULLable in local DB

ChasNelson1990 committed:
feat: refactor BAHIS3 sync functions

ChasNelson1990 committed:
wip!: move to Enketo for form rendering

ChasNelson1990 committed:
feat: add local draft and cloud submission tables to the database

ChasNelson1990 committed:
feat: move to winston for electron-side logging too

ChasNelson1990 committed:
feat: completely rewrite electron-side sign in

ChasNelson1990 committed:
feat: remove unused fetchUserInfo handle

ChasNelson1990 committed:
feat: improve read / update local DB handles plus get / post from kobo functions

ChasNelson1990 committed:
feat: ... whoops! I can't pick this apart because it completely refactors the react app to be much simpler with the same functionality... aaaand also does major verion upgrades for node and tonnes of dependencies...

ChasNelson1990 committed:
feat: rename bahis2 users table to users2

ChasNelson1990 committed:
feat: rename bahis2 lists table to lists2

ChasNelson1990 committed:
feat: rename bahis2 form_choices table to form_choices2

ChasNelson1990 committed:
feat: rename bahis2 app_log table to app_log2

ChasNelson1990 committed:
feat: rename bahis2 app table to app2

ChasNelson1990 committed:
feat: rename bahis2 data table to data2

ChasNelson1990 committed:
feat: move last time sync to footer and add network notifier

ChasNelson1990 committed:
feat: minor additions to theme

ChasNelson1990 committed:
feat: add electron window buttons for manually syncing app data or enforcing an update

Note: update button works but doesn't throw a message dialog

ChasNelson1990 committed:
fix: correct typo on KC API env variable

ChasNelson1990 committed:
fix: correct expected repsonse and upload data type for submitting forms to KoboToolbox API

ChasNelson1990 committed:
feat: separate the draft list out as it's own component

ChasNelson1990 committed:
fix: make it so all three different submitted / editable scenarios that Form could encounter are handled

ChasNelson1990 committed:
fix: use then/catch logic on the promise of form XML data when filling out previous data

ChasNelson1990 committed:
feat: refactor Header to support the new drafts process

ChasNelson1990 committed:
feat: make draftList component work

ChasNelson1990 committed:
feat: make iFrames work

ChasNelson1990 committed:
feat: refactor List component (WIP)

ChasNelson1990 committed:
feat: read user admin regions from DB

ChasNelson1990 committed:
fix: remove unused scripts

ChasNelson1990 committed:
feat: add read app version

ChasNelson1990 committed:
feat: minor improvements to EnketoForm

ChasNelson1990 committed:
feat: minor improvement to Error Page text

ChasNelson1990 committed:
wip: workflows

ChasNelson1990 committed:
feat: remove references to bahis 2 form table

ChasNelson1990 committed:
feat: remove references to bahis 2 list table

ChasNelson1990 committed:
feat: remove references to bahis 2 data table

ChasNelson1990 committed:
feat: remove references to bahis 2 form_choices table

ChasNelson1990 committed:
feat: remove references to bahis 2 app_log table

ChasNelson1990 committed:
feat: remove references to bahis 2 app table

ChasNelson1990 committed:
feat: simplify references to bahis 2 users table

ChasNelson1990 committed:
feat: remove unneeded lodash dependency

ChasNelson1990 committed:
feat: completely remove bahis2 databases and sync

ChasNelson1990 committed:
feat: introduce slightly hacky but working solution for already filled out forms

ChasNelson1990 committed:
fix: make sure we get admin regions as names for form defaults

ChasNelson1990 committed:
feat: improved date support

ChasNelson1990 committed:
feat: add ability to delete drafts

ChasNelson1990 committed:
feat: move to offline fonts

ChasNelson1990 committed:
feat: add workflows to lists

ChasNelson1990 committed:
fix: correct prettier globbing

ChasNelson1990 committed:
fix: correct lint react failures

ChasNelson1990 committed:
fix: remove unneccesary exports

ChasNelson1990 committed:
fix: correct lint electron failures

shaarful commented on electron/main.ts:
```diff
@@ -249,1649 +252,323 @@ const template: Electron.MenuItemConstructorOptions[] = [
-    try {
-        const fetchedRows = db.prepare(queryString).all();
+    const SIGN_IN_ENDPOINT = `${BAHIS2_SERVER_URL}/bhmodule/app-user-verify/`;
+    const current_user = db.prepare('SELECT * from users2 limit 1').get();
```
@ChasNelson1990 Why here table is `users2`

ChasNelson1990 replied:
Because it is still the original BAHIS 2 user table. I did refine it so it should be ready for BAHIS 3 auth but for now it's still the old system.

shaarful commented on electron/localDB.ts:
```diff
@@ -0,0 +1,175 @@
+};
+
+export const createUserInLocalDatabase = async (data, userData, db) => {
+    const insertStmt = db.prepare(`INSERT INTO user (username, password, name, role, upazila) VALUES (?, ?, ?, ?, ?)`);
```
what is the user table name? maybe `users`

ChasNelson1990 replied:
Ah, okay, sorry, it looks like I did migrate everything over to the new simpler BAHIS 3 `user` table but must have left one or two references to the `users2` table. Please repoint all `users2` references to `user` (and check it works :rofl: )

shaarful committed:
Fix users table (#96)

* fix users table name error

* fix: font size

---------

Co-authored-by: Sharful Islam <34204530+sharful09@users.noreply.github.com>

shaarful committed:
Fix bettersqlite error (#98)

* fix users table name error

* fix: font size

* fix: better sqlite errors

fix bahis installation logo
remove .env file from repo

---------

Co-authored-by: Sharful Islam <34204530+sharful09@users.noreply.github.com>

shaarful committed:
format fix

shaarful committed:
add existing user name

shaarful committed:
fix: enketo field duplication

ChasNelson1990 committed:
feat: improve Enketo loadErrors handling (#97)

shaarful committed:
fix: remove unnecessary comments

shaarful committed:
fix: win build errors

fixed the windows build errors.
fixed taxonomy path

shaarful committed:
Merge branch 'fix-bahis3-bugs' into bahis3

shaarful committed:
feat: add authentication (#100)

* feat: add authentication via token

get api token from kobo while login and save into database
use the token for next comunication
changed the user model
fix log for error consoling

* feat: add logout system

* remove key from .env

shaarful committed:
Fix bug (#101)

* feat: add authentication via token

get api token from kobo while login and save into database
use the token for next comunication
changed the user model
fix log for error consoling

* feat: add logout system

* fix: draft count bug

draft count did not updated after submitted a form. now it has been
fixed.
redux added

* fix: counting bugs

* feat: Automatic logout every 7 days

* fix: retry failed request

The app will retry a failed request up to 10 times.
Data will be inserted into the database in a single transaction, improving the app's speed.
app will now pull data starting from the last fetched record onward.

* feat: add notification and status massage

add new notification and status of current state of app to make more
interactive

* Update .env

* feat: migrate to esm add tailwind imporved styling

migrate commonjs to es module
add tailwindcss for styling
improved theme and modify some text
add bahis logo and text to app bar

* fix: delete env file

* fix: some styling issues

* fix: change notification system

now notification stackable

shaarful committed:
fix: prettier

shaarful committed:
fix: env file

shaarful committed:
fix: env file

shaarful committed:
Fix bugs (#102)

* fix: wrong message

* fix: not showing submit button

* fix: list module not showing value

* fix: list module not rendering row

* fix: last logged in

* fix: delete remove modules and few adjustment

* fix: max stack and others

* fix: list not showing for no start date found

* fix: notification and status improvement

* feat: add info button
